db/db_compaction_test.cc

```diff
@@ -10042,6 +10042,170 @@ TEST_F(DBCompactionTest, VerifyRecordCount) {
       "processed.";
   ASSERT_TRUE(std::strstr(s.getState(), expect));
 }
+
+TEST_F(DBCompactionTest, ReleaseCompactionDuringManifestWrite) {
+  // Tests the fix for issue #10257.
+  // Compactions are released in LogAndApply() so that picking a compaction
+  // from the new Version won't see these compactions as registered.
+  Options options = CurrentOptions();
+  options.compaction_style = kCompactionStyleLevel;
+  // Make sure we can run multiple compactions at the same time.
+  env_->SetBackgroundThreads(3, Env::Priority::LOW);
+  env_->SetBackgroundThreads(3, Env::Priority::BOTTOM);
+  options.max_background_compactions = 3;
+  options.num_levels = 4;
+  DestroyAndReopen(options);
+  Random rnd(301);
+
+  // Construct the following LSM
+  // L1:           [K10-K11]
+  // L2:  [K1-K2]  [K10]          [k100-k101]
+  // L3: [K1]                     [k100]
+  // We will have 3 threads to run 3 manual compactions.
+  // T0 runs CompactRange(K100, K101)
+  // T1 runs CompactRange(K1, K2)
+  // T2 runs CompactRange(K10, K11)
+  // T1 and T2 will wait for T0 to write to MANIFEST and will be in
+  // same write group.
+  //
+  // A rough timeline of events:
+  // - T0 enters LogAndApply() to write compaction result to manifest
+  // - T0 unlocks mutex when writing to manifest
+  // - T1 starts compaction
+  // - T1 enters LogAndApply(), becomes leader of the next write group,
+  //   and waits for T0 to finish
+  // - T2 starts compaction
+  // - T2 enters LogAndApply(), joins T1's write group, wait for T1 to finish
+  // - Now T0 resumes, finishes compaction
+  // - T1, T2 groups commits their manifest writes, new version creation,
+  //   and finish compaction
```

Comment:
It's a good test case. Does it matter what order the compactions happen? Assuming not, how about the following approach?

- Launch three threads to run CompactRange()
- The first one to reach "VersionSet::LogAndApply:WriteManifestStart" waits for the remaining two to pass "VersionSet::LogAndApply:BeforeWriterWaiting"
- The second one to reach "DBImpl::BackgroundCompaction:AfterCompaction" asserts no compactions remain registered

Reply:
Thanks for the suggestion. I've simplified the tests. The complexity is partially due to my original try to also show that an L0->L1 compaction cannot be scheduled.